test/tools/cmap_spec_runner.ts

```diff
@@ -0,0 +1,336 @@
+ */
+const compareInputToSpec = (input, expected) => {
+  if (expected === 42) {
+    expect(input).to.be.ok; // not null or undefined
+    return;
+  }
+
+  if (Array.isArray(expected)) {
+    expect(input).to.be.an('array');
+    expected.forEach((expectedValue, index) => {
+      compareInputToSpec(input[index], expectedValue);
+    });
+    return;
+  }
+
+  if (expected && typeof expected === 'object') {
+    for (const [expectedPropName, expectedValue] of Object.entries(expected)) {
+      expect(input).to.have.property(expectedPropName);
+      compareInputToSpec(input[expectedPropName], expectedValue);
+    }
+    return;
+  }
+
+  expect(input).to.equal(expected);
+};
+
+const getTestOpDefinitions = (threadContext: ThreadContext) => ({
+  checkOut: async function (op) {
+    const connection: Connection = await util
+      .promisify(ConnectionPool.prototype.checkOut)
+      .call(threadContext.pool);
+    if (op.label != null) {
+      threadContext.connections.set(op.label, connection);
+    } else {
+      threadContext.orphans.add(connection);
+    }
+  },
+  checkIn: function (op) {
+    const connection = threadContext.connections.get(op.connection);
+    threadContext.connections.delete(op.connection);
+
+    if (!connection) {
+      throw new Error(`Attempted to release non-existient connection ${op.connection}`);
+    }
+
+    return threadContext.pool.checkIn(connection);
+  },
+  clear: function () {
+    return threadContext.pool.clear();
+  },
+  close: function () {
+    return util.promisify(ConnectionPool.prototype.close).call(threadContext.pool);
+  },
+  wait: async function (options) {
+    const ms = options.ms;
+    return sleep(ms);
+  },
+  start: function (options) {
+    const target = options.target;
+    const thread = threadContext.getThread(target);
+    thread.start();
+  },
+  waitForThread: async function (options): Promise<void> {
+    const name = options.name;
+    const target = options.target;
+
+    const threadObj = threadContext.threads.get(target);
+
+    if (!threadObj) {
+      throw new Error(`Attempted to run op ${name} on non-existent thread ${target}`);
+    }
+
+    await threadObj.finish();
+  },
+  waitForEvent: function (options): Promise<void> {
+    const event = options.event;
+    const count = options.count;
+    return new Promise(resolve => {
+      function run() {
+        if (threadContext.poolEvents.filter(ev => getEventType(ev) === event).length >= count) {
+          return resolve();
+        }
+
+        threadContext.poolEventsEventEmitter.once('poolEvent', run);
+      }
+      run();
+    });
+  }
+});
+
+export class ThreadContext {
+  pool: ConnectionPool;
+  threads: Map<any, Thread>;
+  connections: Map<string, Connection>;
+  orphans: Set<Connection>;
+  poolEvents = [];
+  poolEventsEventEmitter = new EventEmitter();
+  hostAddress;
+  supportedOperations;
```

Comment:
```suggestion
  supportedOperations: ReturnType<typeof getTestOpDefinitions>;
```

It can start here but then lead to typing the indexing of `_knownCommands` in thread, I know we have a runtime check but the more we try to strictly type the better early warning one gets when they start modifying this code for the next runner update. (Within reason, we've already improved greatly I think the type improvements I'm calling out are not blockers.)

Reply:
I'll see what I can do

Reply:
made just this change